src/sentry/api/event_search.py

```diff
@@ -1633,31 +1643,68 @@ def parse_function(field, match=None):
     )
 
 
+def count_required_arguments(function):
+    n = 0
+    for arg in function["args"]:
+        if arg.has_default:
+            break
+        n += 1
+    return n
+
+
+def count_total_arguments(function):
+    return len(function["args"])
+
+
+def validate_argument_count(field, function, arguments):
+    """
+    Validate the number of required arguments the function defines against
+    provided arguments. Raise an exception if there is a mismatch in the
+    number of arguments. Do not return any values.
+
+    There are 4 cases:
+        1. provided # of arguments != required # of arguments AND provided # of arguments != total # of arguments (bad, raise an error)
+        2. provided # of arguments < required # of arguments (bad, raise an error)
+        3. provided # of arguments > total # of arguments (bad, raise an error)
+        4. required # of arguments <= provided # of arguments <= total # of arguments (good, pass the validation)
+    """
+    args_count = len(arguments)
+    total_args_count = count_total_arguments(function)
+    if args_count != total_args_count:
+        required_args_count = count_required_arguments(function)
+        if required_args_count == total_args_count:
+            raise InvalidSearchQuery(
+                u"{}: expected {:g} arguments".format(field, len(function["args"]))
+            )
+        elif args_count < required_args_count:
+            raise InvalidSearchQuery(
+                u"{}: expected at least {:g} arguments".format(field, required_args_count)
+            )
+        elif args_count > total_args_count:
+            raise InvalidSearchQuery(
+                u"{}: expected at most {:g} arguments".format(field, total_args_count)
+            )
+
+
 def resolve_function(field, match=None, params=None):
     function, columns = parse_function(field, match)
     function = FUNCTIONS[match.group("function")]
 
-    # Some functions can optionally take no parameters (epm(), eps()). In that case use the
-    # passed in params to create a default argument if necessary.
-    used_default = False
-    if len(columns) == 0 and len(function["args"]) == 1:
+    validate_argument_count(field, function, columns)
+
+    columns_with_defaults = [col for col in columns]
+    for argument in function["args"][len(columns) :]:
+        assert argument.has_default
```

Comment:
What sort of error does this return to the user if the assert fails? Also would this ever fail?

Reply:
I meant to take this out but this should never fail in production but acts as a safe guard during development to stop someone from providing too few arguments. There is also an unit test in place to ensure that functions are defined correctly.